plugin/plugins/kafka/service/service.go

```diff
@@ -0,0 +1,405 @@
+	key := &MsgKey{
+		BlockHeight: kss.currentBlockNumber,
+		Event:       MsgKey_Event(event),
+		EventId:     eventId,
+		EventType: 	 MsgKey_EventType(REQUEST),
+		EventTypeId: eventTypeId,
+	}
+
+	// write req
+	if err := kss.writeAsJsonToKafka(ctx, EndBlockReqTopic, key, &req); err != nil {
+		return err
+	}
+
+	// write state changes
+	if err := kss.writeStateChange(ctx, event, eventId); err != nil {
+		return err
+	}
+
+	// write res
+	key.EventType = MsgKey_EventType(RESPONSE)
+	if err := kss.writeAsJsonToKafka(ctx, EndBlockResTopic, key, &res); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// HaltAppOnDeliveryError whether or not to halt the application when delivery of massages fails
+// in ListenBeginBlock, ListenEndBlock, ListenDeliverTx. Setting this to `false` will give fire-and-forget semantics.
+// When `true`, the app will gracefully halt and stop the running node. Uncommitted blocks will
+// be replayed to all listeners when the node restarts and all successful listeners that received data
+// prior to the halt will receive duplicate data.
+func (kss *KafkaStreamingService) HaltAppOnDeliveryError() bool {
+	return kss.haltAppOnDeliveryError
+}
+
+// Stream spins up a goroutine select loop which awaits length-prefixed binary encoded KV pairs and caches them in the order they were received
+// Do we need this and an intermediate writer? We could just write directly to the buffer on calls to Write
+// But then we don't support a Stream interface, which could be needed for other types of streamers
+func (kss *KafkaStreamingService) Stream(wg *sync.WaitGroup) error {
+	if kss.quitChan != nil {
+		return errors.New("`Stream` has already been called. The stream needs to be closed before it can be started again")
+	}
+	kss.quitChan = make(chan struct{})
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-kss.quitChan:
+				return
+			case by := <-kss.srcChan:
+				kss.stateCacheLock.Lock()
+				kss.stateCache = append(kss.stateCache, by)
+				kss.stateCacheLock.Unlock()
+			}
+		}
+	}()
+	return nil
+}
+
+// Close satisfies the io.Closer interface
+func (kss *KafkaStreamingService) Close() error {
+	kss.producer.Flush(kss.flushTimeoutMs)
+	close(kss.quitChan)
+	close(kss.deliveryChan)
+	kss.producer.Close()
+	return nil
+}
+
+func (kss *KafkaStreamingService) writeStateChange(ctx sdk.Context, event int64, eventId int64) error {
+	// write all state changes cached for this stage to Kafka
+	kss.stateCacheLock.Lock()
+	kvPair := new(types.StoreKVPair)
+	for i, stateChange := range kss.stateCache {
+		key := &MsgKey{
+			BlockHeight: kss.currentBlockNumber,
+			Event:       MsgKey_Event(event),
+			EventId:     eventId,
+			EventType:   MsgKey_EventType(STATE_CHANGE),
+			EventTypeId: int64(i + 1),
+		}
+		if err := kss.codec.UnmarshalLengthPrefixed(stateChange, kvPair); err != nil {
+			return err
+		}
+		if err := kss.writeAsJsonToKafka(ctx, StateChangeTopic, key, kvPair); err != nil {
+			return err
+		}
```

Comment:
If either of these error conditions are triggered, kss.stateCacheLock is never unlocked, resulting in deadlock.

Reply:
fixed.